### ISSUE TITLE
Inconsistent behavior of Reaction-Gene connections in SBML reading

### ISSUE THREAD

**Issue by cdiener** [opencobra/cobrapy#1185]
#### Problem description

SBML FBC allows genes to be in a model even if there are no GPRs. However, reading such models creates some odd behaviors with cobrapy and breaks round-trips. This behavior was introduced in #1172. 

It's a bit undefined behavior here since FBC does not have a Reaction-Gene connection. There are only genes for the entire model and GPRs. However, the behavior does look a bit odd. As far as I can tell it does not delete genes from `model.genes` for now, which is good as this would be more problematic.

#### Code Sample

```python
Python 3.9.10 (main, Feb 22 2022, 13:54:07) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cobra.io import read_sbml_model, write_sbml_model

In [2]: from cobra.test import create_test_model

In [3]: mod = create_test_model("textbook")

In [4]: for r in mod.reactions:
   ...:     r.gpr.body = None
   ...: 

In [5]: write_sbml_model(mod, "textbook_no_gprs.xml")

In [6]: mod.reactions[0].genes
Out[6]: frozenset({<Gene b0351 at 0x7f218325d790>, <Gene b1241 at 0x7f218325d700>})

In [7]: read = read_sbml_model("textbook_no_gprs.xml")

In [8]: read.reactions[0].genes
Out[8]: frozenset()
```

The written model will pass the SBML validator without errors.

### Context

<details>

```
System Information
==================
OS                  Linux
OS-release 5.16.0-4-amd64
Python             3.9.10

Package Versions
================
appdirs~=1.4                            not installed
black                                          22.1.0
bumpversion                                     0.6.0
cobra                                          0.24.0
depinfo~=1.7                            not installed
diskcache~=5.0                          not installed
future                                         0.18.2
httpx~=0.14                             not installed
importlib_resources                             3.3.0
isort; extra == "development"           not installed
numpy~=1.13                             not installed
optlang~=1.5                            not installed
pandas~=1.0                             not installed
pip                                            22.0.4
pydantic~=1.6                           not installed
python-libsbml~=5.19                    not installed
rich                                           11.2.0
ruamel.yaml~=0.16                       not installed
scipy                                           1.8.0
setuptools                                     60.9.3
swiglpk                       4.65.1+6.gd955ae9.dirty
tox                                            3.23.0
wheel                                          0.37.1
```

</details>

**Comment by akaviaLab** [comment]
It wasn't exactly introduced only in #1172
If you look at the repository before I started my changes [reaction.py](https://github.com/opencobra/cobrapy/blob/0bbad14ee36406ea14c06f44e9cdd216182b945a/src/cobra/core/reaction.py#L418) you can see that the setter for gene_reaction_rule basically tries to figure out which are the new genes, and then deletes genes which aren't that from the reaction.

Hmm. I'm thinking about this a bit.

Could we add a flag to the update_genes_from_gpr() like delete_genes, with default False?
Some calls will change this to True and delete genes:
1) When trying to manually set the gene_reaction_rules
2) When setting the GPR. I think that if you're manually setting the GPR, you should know what you're doing. Similarly when you're manually setting the gene_reaction_rules. 
3) When renaming genes
4) When replacing genes

What do you think about this suggestion and these situations?

**Comment by akaviaLab** [comment]
Or I could have it be default False,
when renaming/replacing give a list/set of genes you can remove
in some situations, give True, which means delete all of them.

**Comment by cdiener** [comment]
I think the issue is more the inconsistent behavior. I think it's fine to remove genes from the *reaction* if the GPR does not contain them, but they should remain in `model.genes` unless you specify that you want them gone there. But `remove_genes` will also delete them from the model and is now called when updating the GPR and that should not happen. And this behavior was not there before #1172 because it used `g._reaction.remove(self)` instead of `remove_genes`.

**Comment by cdiener** [comment]
So in summary I agree with cases 3 and 4 but disagree with 1 and 2 since those are incompatible with the SBML spec which allows genes to exist in a model even if they don't appear in a GPR. Also makes sense since you may know a gene is present in an organism without being able to match it to a biochemical reaction.

Edit: Another case where that becomes problematic is when a gene would participate in two enzymes/reactions. In that case, removing it from one would remove it from the other as well.

**Comment by akaviaLab** [comment]
Hi.
Trying to summarize what changes you'd like.
First, under no circumstances should updating the GPR or genes based on GPR remove genes from the *model*. Correct?

Second, there is the discussion of updating/removing genes from the reaction. We both agree that when renaming/removing genes from the model, they should be removed from the reaction. Correct? Should other genes be removed as well?

Let's say a reaction has GPR A & B, and reaction.genes= {A, B, C} without C being in GPR.
Rename A -> A2, GPR A & B -> A2 & B, reaction.genes = {A2, B, C}? reaction.genes = {A2, B}? Whatever the answer is, it will apply to removing genes from model.

Updating GPR/gene_reaction_rule manually. To me, it makes sense that if you're manually updating GPR/gene_reaction_rule, and you don't explicitly include genes, they should be removed from the reaction. Otherwise, you could have added them in.
So
rxn has GPR=(A & B), genes = {A, B, C, D}
assign GPR=(A & D) genes {A, B, C, D} -> {A, D}

I think this logic contradicts with the SBML standard, if I understood you correctly. If so, what do you think should happen?
Should genes look like {A, B, C, D} or {A, C, D}? or something else? 

I'm not sure what this section means
"Edit: Another case where that becomes problematic is when a gene would participate in two enzymes/reactions. In that case, removing it from one would remove it from the other as well."
If you're updating the GPR of one reaction, the other reaction should not be updated (I'll fix GPR not to remove genes from model).
If the model is having genes removed, it will be removed from both reactions.
Can you please illustrate it?

**Comment by cdiener** [comment]
Yeah sorry that was confusing because I wasn't aware the update deletes genes from the model. This issue was more about the timing of the update operation which is too delayed. 

So for the issue discussed now the main idea is that Reaction.genes would be in sync with the GPR but model.genes does not have to be.

> Hi. Trying to summarize what changes you'd like. First, under no circumstances should updating the GPR or genes based on GPR remove genes from the _model_. Correct?

Correct.

> Second, there is the discussion of updating/removing genes from the reaction. We both agree that when renaming/removing genes from the model, they should be removed from the reaction. Correct? Should other genes be removed as well?

Genes explicitly removed with remove_gene will be removed from the model, yes. And renamed genes will also be renamed in the model (so the old name/gene will be deleted and replaced with the new). Other genes would be removed from the *reaction* but not from the *model*.

> Let's say a reaction has GPR A & B, and reaction.genes= {A, B, C} without C being in GPR. Rename A -> A2, GPR A & B -> A2 & B, reaction.genes = {A2, B, C}? reaction.genes = {A2, B}? Whatever the answer is, it will apply to removing genes from model.

`reaction.genes = {A2, B}` but `model.genes = {A2, B, C}` in that case. You have stated that C does not participate in the enzyme for the reaction, however the gene is still part of the organism.

> Updating GPR/gene_reaction_rule manually. To me, it makes sense that if you're manually updating GPR/gene_reaction_rule, and you don't explicitly include genes, they should be removed from the reaction. Otherwise, you could have added them in. So rxn has GPR=(A & B), genes = {A, B, C, D} assign GPR=(A & D) genes {A, B, C, D} -> {A, D}

Agreed for the *reaction*. I just want them to stay in the *model*.

> I think this logic contradicts with the SBML standard, if I understood you correctly. If so, what do you think should happen? Should genes look like {A, B, C, D} or {A, C, D}? or something else?

SBML has no notion of reaction-gene relationships. Genes belong to a model and can appear in reaction GPRs but reactions have no genes. So on the *reaction* level there is no limitation imposed by SBML, but there is on the *model* level.

> I'm not sure what this section means "Edit: Another case where that becomes problematic is when a gene would participate in two enzymes/reactions. In that case, removing it from one would remove it from the other as well." If you're updating the GPR of one reaction, the other reaction should not be updated (I'll fix GPR not to remove genes from model). If the model is having genes removed, it will be removed from both reactions. Can you please illustrate it?

Again that is related to deleting genes from the model. There is nothing that prohibits a gene to be part of 2 or more GPRs. If altering a single GPR would delete it from the model that would automatically affect all other GPRs and reactions that use that gene as a side effect.

---

So this issue here was more the timing of the update. I would want the removal on the reaction level to happen immediately upon modification of the GPR and not delayed in some copy operation etc. The example was probably not well chosen, I should have set the GPR to an empty string or something but it looks like that would have deleted all genes then?

**Comment by cdiener** [comment]
Also this my fault for not looking at the initial PRs so sorry for that. If you'd like I could also make a PR with the changes I'd like and you can review it (or reject it). Maybe that is easier.

**Comment by akaviaLab** [comment]
I think I addressed everything in #1187 
I made it one slightly big PR since I had like three branches open all dealing with variants of the same thing.

**Comment by akaviaLab** [comment]
I thought it would be closed now that #1187 is merged.
Need to close.
